rcmgr.go

```diff
@@ -0,0 +1,441 @@
+package rcmgr
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+type ResourceManager struct {
+	limits Limiter
+
+	system    *SystemScope
+	transient *TransientScope
+
+	mx    sync.Mutex
+	svc   map[string]*ServiceScope
+	proto map[protocol.ID]*ProtocolScope
+	peer  map[peer.ID]*PeerScope
+
+	cancelCtx context.Context
+	cancel    func()
+	wg        sync.WaitGroup
+}
+
+var _ network.ResourceManager = (*ResourceManager)(nil)
+
+type SystemScope struct {
+	*ResourceScope
+}
+
+var _ network.ResourceScope = (*SystemScope)(nil)
+
+type TransientScope struct {
+	*ResourceScope
+
+	system *SystemScope
+}
+
+var _ network.ResourceScope = (*TransientScope)(nil)
+
+type ServiceScope struct {
+	*ResourceScope
+
+	name   string
+	system *SystemScope
+}
+
+var _ network.ServiceScope = (*ServiceScope)(nil)
+
+type ProtocolScope struct {
+	*ResourceScope
+
+	proto  protocol.ID
+	system *SystemScope
+}
+
+var _ network.ProtocolScope = (*ProtocolScope)(nil)
+
+type PeerScope struct {
+	*ResourceScope
+
+	peer  peer.ID
+	rcmgr *ResourceManager
+}
+
+var _ network.PeerScope = (*PeerScope)(nil)
+
+type ConnectionScope struct {
+	*ResourceScope
+
+	dir   network.Direction
+	usefd bool
+	rcmgr *ResourceManager
+	peer  *PeerScope
+}
+
+var _ network.ConnectionScope = (*ConnectionScope)(nil)
+var _ network.ConnectionManagementScope = (*ConnectionScope)(nil)
+
+type StreamScope struct {
+	*ResourceScope
+
+	dir   network.Direction
+	rcmgr *ResourceManager
+	peer  *PeerScope
+	svc   *ServiceScope
+	proto *ProtocolScope
+}
+
+var _ network.StreamScope = (*StreamScope)(nil)
+var _ network.StreamManagementScope = (*StreamScope)(nil)
+
+func NewResourceManager(limits Limiter) *ResourceManager {
```

Comment:
```suggestion
func NewResourceManager(limits Limiter) network.ResourceManager {
```

Reply:
wait, we may want to export the rcmgr type with a wider interface.

For example we could add a refcounting interface to allow sharing of rcmgrs between multiple hosts.
Then again this could be a wrapper type.

Reply:
done.